node/src/components/fetcher/trie_fetcher.rs

```diff
@@ -0,0 +1,342 @@
+    sender: Option<I>,
+}
+
+impl<I> PartialChunks<I>
+where
+    I: Debug + Eq,
+{
+    fn missing_chunk(&self, count: u64) -> Option<u64> {
+        (0..count).find(|idx| !self.chunks.contains_key(idx))
+    }
+
+    fn mutate_sender(&mut self, sender: Option<I>) {
+        let old_sender = self.sender.take();
+        self.sender = old_sender.or(sender);
+    }
+
+    fn assemble_chunks(&self, count: u64) -> Result<Trie<Key, StoredValue>, bytesrepr::Error> {
+        let data: Vec<u8> = (0..count)
+            .filter_map(|index| self.chunks.get(&index))
+            .flat_map(|chunk| chunk.chunk())
+            .copied()
+            .collect();
+        bytesrepr::deserialize(data)
+    }
+}
+
+#[derive(DataSize, Debug)]
+pub(crate) struct TrieFetcher<I>
+where
+    I: Debug + Eq,
+{
+    partial_chunks: HashMap<Digest, PartialChunks<I>>,
+}
+
+#[derive(DataSize, Debug, From)]
+pub(crate) enum Event<I>
+where
+    I: Debug + Eq,
+{
+    #[from]
+    Request(TrieFetcherRequest<I>),
+    TrieOrChunkFetched {
+        id: TrieOrChunkedDataId,
+        fetch_result: FetchResult<TrieOrChunkedData, I>,
+    },
+}
+
+impl<I> fmt::Display for Event<I>
+where
+    I: Debug + Eq,
+{
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Event::Request(_) => write!(f, "trie fetcher request"),
+            Event::TrieOrChunkFetched { id, .. } => write!(f, "trie or chunk {} fetched", id),
+        }
+    }
+}
+
+fn into_response<I>(trie: Box<Trie<Key, StoredValue>>, sender: Option<I>) -> TrieFetcherResult<I>
+where
+    I: Debug + Eq,
+{
+    match sender {
+        Some(peer) => Ok(FetchedData::FromPeer { item: trie, peer }),
+        None => Ok(FetchedData::FromStorage { item: trie }),
+    }
+}
+
+impl<I> TrieFetcher<I>
+where
+    I: Debug + Clone + Send + Eq + 'static,
+{
+    pub(crate) fn new() -> Self {
+        TrieFetcher {
+            partial_chunks: Default::default(),
+        }
+    }
+
+    fn consume_trie_or_chunk<REv>(
+        &mut self,
+        effect_builder: EffectBuilder<REv>,
+        sender: Option<I>,
+        trie_or_chunk: TrieOrChunkedData,
+    ) -> Effects<Event<I>>
+    where
+        REv: ReactorEventT<TrieOrChunkedData> + From<FetcherRequest<I, TrieOrChunkedData>>,
+    {
+        let TrieOrChunkedDataId(_index, hash) = trie_or_chunk.id();
+        match trie_or_chunk {
+            TrieOrChunkedData::Trie(trie) => match self.partial_chunks.remove(&hash) {
+                None => {
+                    error!(%hash, "fetched a trie we didn't request!");
+                    Effects::new()
+                }
+                Some(partial_chunks) => {
+                    debug!(%hash, "got a full trie");
+                    partial_chunks
+                        .responder
+                        .respond(into_response(trie, sender))
+                        .ignore()
+                }
+            },
+            TrieOrChunkedData::ChunkWithProof(chunk) => {
```

Comment:
Should the `chunk.id() == index` from the above?

Reply:
The index gets calculated from the chunk, anyway, so it will always be equal.